usaspending_api/recipient/delta_models/recipient_lookup.py

```diff
@@ -1,30 +1,583 @@
+            COALESCE(fpds.legal_entity_congressional, fabs.legal_entity_congressional) AS congressional_district,
+            COALESCE(fpds.legal_entity_address_line1, fabs.legal_entity_address_line1) AS address_line_1,
+            COALESCE(fpds.legal_entity_address_line2, fabs.legal_entity_address_line1) AS address_line_2,
+            COALESCE(fpds.legal_entity_country_code, fabs.legal_entity_country_code) AS country_code,
+            tn.action_date,
+            CASE
+                WHEN tn.is_fpds = TRUE THEN CAST('fpds' AS STRING)
+                ELSE CAST('fabs' AS STRING)
+            END AS source
+        FROM raw.transaction_normalized AS tn
+        LEFT OUTER JOIN raw.transaction_fpds AS fpds ON (tn.id = fpds.transaction_id)
+        LEFT OUTER JOIN raw.transaction_fabs AS fabs ON (tn.id = fabs.transaction_id)
+        WHERE tn.action_date >= '2007-10-01'
+        ORDER BY tn.action_date DESC
+    )
+    """,
+    # -----
+    # Creation of the temporary table that is used to stage and merge updates to recipient_lookup
+    # -----
+    rf"""
+    CREATE OR REPLACE TABLE temp.temporary_restock_recipient_lookup (
+        {", ".join([f'{key} {val}' for key, val in TEMP_RECIPIENT_LOOKUP_COLUMNS.items()])}
+    )
+    USING DELTA
+    LOCATION 's3a://{{SPARK_S3_BUCKET}}/{{DELTA_LAKE_S3_PATH}}/temp/temporary_restock_recipient_lookup'
+    """,
+    # -----
+    # Populate the temporary_restock_recipient_lookup table
+    # -----
+    r"""
+    WITH latest_duns_sam AS (
+        SELECT
+            1 AS priority,
+            REGEXP_REPLACE(
+                MD5(UPPER(
+                    CASE WHEN uei IS NOT NULL THEN CONCAT('uei-', uei)
+                    ELSE CONCAT('duns-', awardee_or_recipient_uniqu) END
+                )),
+                '^(\.{{8}})(\.{{4}})(\.{{4}})(\.{{4}})(\.{{12}})$',
+                '\$1-\$2-\$3-\$4-\$5'
+            ) AS recipient_hash,
+            REGEXP_REPLACE(
+                MD5(UPPER(CONCAT('duns-', awardee_or_recipient_uniqu))),
+                '^(\.{{8}})(\.{{4}})(\.{{4}})(\.{{4}})(\.{{12}})$',
+                '\$1-\$2-\$3-\$4-\$5'
+            ) AS duns_recipient_hash,
+            UPPER(legal_business_name) AS legal_business_name,
+            awardee_or_recipient_uniqu AS duns,
+            uei,
+            'sam' AS source,
+            address_line_1,
+            address_line_2,
+            business_types_codes,
+            city,
+            congressional_district,
+            country_code,
+            ultimate_parent_unique_ide AS parent_duns,
+            UPPER(ultimate_parent_legal_enti) AS parent_legal_business_name,
+            ultimate_parent_uei AS parent_uei,
+            state,
+            zip4,
+            zip AS zip5,
+            update_date,
+            NULL AS alternate_names,
+            ROW_NUMBER() OVER (PARTITION BY uei, awardee_or_recipient_uniqu ORDER BY update_date DESC NULLS LAST) AS row_num
```

Comment:
In this implementation and the prototype code there is a CTE below
```
union_all_priority AS (
        SELECT
            *,
            ROW_NUMBER() OVER (PARTITION BY recipient_hash ORDER BY priority ASC) as row_num_union
        FROM union_all
    )
```

This would limit the value based on the hash which in the case of sam does not take into account the `legal_business_name`. Since it doesn't take into account the name it seems that it is wasted effort to try and partition by the name as well since two recipients with the same UEI and different names would get condensed into a single recipient anyways.

As for the COALESCE on UEI and DUNS I don't think it makes any difference. I have not tried with the COALESCE, but I have seen the data match without it.